rfd/0092-error-message-guidelines.md

```diff
@@ -0,0 +1,552 @@
+---
+authors: Jeff Anderson <jeff@goteleport.com>
+state: draft
+---
+
+# RFD 0092 - Error Message Guidelines
+
+As more companies deploy Teleport, more and more less-technical users whose eyes glaze over at highly technical error messages and stacktraces will be using our products.
+
+Teleport can adopt some opinions on how its error messages and stacktraces are communicated to both the Teleport administrators and end users.
+
+While much of this document is hopefully describing things that may already be goals, I wrote it based on the list of "case studies" further down. These are actual stacktraces and error messages that customers have asked about through support channels. I've tried to include at least one reference for each one, but I didn't do an exhaustive search to enumerate all occurrences.
+
+## Suggested Opinions to adopt
+
+After going through several cases studies, the following guidelines should be taken into account when determining error, logging, or stacktrace behavior.
+
+### Unify the logging strategy
+
+There are currently subtle differences in what logs and error messages look like. Sometimes, the stack trace is escaped and prints a literal `\n` instead of a newline. Sometimes it prints a multiline stacktrace. Sometimes html escaped entities are used, like `&gt;` or `&#39;`, which clog up the readibility.
+
+This isn't a huge concern, but the subtle variations that we do can get in the way from time to time.
+
+### prefer short messages instead of multiline messages
+
+Full Stacktraces should be unexpected and rare. Any error message that is lower than debug should be a shorter version of what might be a multiline stacktrace or error message. Many shorter messages already exist in the product today. This is an example of one of the longer single line messages, but it does a good job getting relevant information across:
```

Comment:
I would prefer we change our logging text formatter rather than avoid `log.WithError`. If we use `WithError`, then the JSON logs will have an `"error": "..."` entry, which is nice. If we instead insert the error string into the log message, then it intermingles with the log message under `"message":  "..."`.

I also think it's more understandable and maintainable to use this approach rather than relying on code reviewers to remember the subtle nature of the current `WithError` behavior. "Always use `WithError` to log errors" is a simpler guideline to follow.

Reply:
I traced it down to this custom log formatting code here:
https://github.com/gravitational/teleport/blob/15f034f40c5256ed02525846da50c21a9a3d1578/lib/utils/formatter.go#L260-L267

My preference would be to encourage use of `WithError` and change our log formatter to have a configurable (opt-in) setting to print stack traces, for example:
```go
func (w *writer) writeError(value interface{}, debugReport bool) {
	switch err := value.(type) {
	case trace.Error:
                 if debugReport {
		    w.WriteString(fmt.Sprintf("[%v]", err.DebugReport()))
                    return
                 }
	}
        w.WriteString(fmt.Sprintf("[%v]", value))
}
```
 That would be the best way to ensure uniformity throughout our logging imo.

Reply:
Adding onto this: If we always use `WithError` then we can modify the formatting for error messages separate from the log message.

For example, I tested a small change to our json formatter that does just that. My change replaces all `"` chars with a backtick in all messages so that the json log doesn't insert escapes, and marshals the trace error as JSON instead of a string:

```
{
  "caller": "cache/cache.go:805",
  "component": "auth:1:cache",
  "error": {
    "message": "invalid origin value `bad-config`, must be one of [defaults config-file dynamic cloud kubernetes]",
    "user_messages": [
      "database `pg`",
      "failed to fetch resource: `db`"
    ]
  },
  "level": "warning",
  "message": "Cache `auth` first init failed, continuing re-init attempts in background.",
  "timestamp": "2023-01-31T14:26:52-08:00"
}
```

Reply:
And regarding `"` string replacement: If we do agree to eliminate escaped quotes in our logs, I'd prefer we have a lint against using `%q` or `"\"thing\""` in format strings rather than doing a string replacement for every log message.